### PR TITLE
Add analysis for extends clause

### PR DESCRIPTION
**Pull request by bioball**
Also, this adds a quickfix for implementing members.

* Insert members and also imports when needed
* Refactor: merge `shortDisplayName` and `name` (they are duplicate properties that mean the same thing)
* Move `ModifierQuickfixTest` to `PklAddModifierQuickFixTest`; make sibling of new test
* Fix add modifier logic when no module header exists

### CODE REVIEW

File: src/test/kotlin/org/pkl/lsp/actions/PklImplementMembersQuickFixTest.kt [hunk 1]
```diff
@@ -0,0 +1,465 @@
...
+          abstract fixed name: String
+        }
+        class Child extends Base {
+          fixed name: String = TODO()
+        }
+        """
+          .trimIndent(),
+    )
+  }
+
```
Comment (stackoverflow) [inline src/test/kotlin/org/pkl/lsp/actions/PklImplementMembersQuickFixTest.kt:161]: The quickfix is adding the property out of the class when there's already something in the body:
```suggestion

  @Test
  fun `quickfix implements member in class with existing body`() {
    checkQuickFix(
      before =
        """
        abstract module Foo

        abstract class Base {
          abstract name: String
        }
        class Child extends Base {
          bar: Int = 5
        }
        """
          .trimIndent(),
      after =
        """
        abstract module Foo

        abstract class Base {
          abstract name: String
        }
        class Child extends Base {
          bar: Int = 5

          name: String = TODO()
        }
        """
          .trimIndent(),
    )
  }
```

File: src/test/kotlin/org/pkl/lsp/actions/PklImplementMembersQuickFixTest.kt [hunk 2]
```diff
@@ -0,0 +1,465 @@
...
+  @Test
```
Comment (stackoverflow) [inline src/test/kotlin/org/pkl/lsp/actions/PklImplementMembersQuickFixTest.kt:162]: The quickfix is fixing properties that already have a default.
```suggestion
  @Test
  fun `quickfix does not reimplement fixed property that has an inherited default`() {
    checkQuickFix(
      before =
        """
        abstract module Foo

        abstract class Base {
          fixed name: String = "default"
          abstract function greet(): String
        }
        class Child extends Base {}
        """
          .trimIndent(),
      after =
        """
        abstract module Foo

        abstract class Base {
          fixed name: String = "default"
          abstract function greet(): String
        }
        class Child extends Base {
          function greet(): String = TODO()
        }
        """
          .trimIndent(),
    )
  }
  
  @Test
```

File: src/test/kotlin/org/pkl/lsp/GoToDefinitionPackagesTest.kt [hunk 3]
```diff
@@ -60,27 +60,27 @@ class GoToDefinitionPackagesTest : LspTestBase() {
...
       }
-    }
-    """
+      """
         .trimIndent()
     )
     val resolved = goToDefinition().single()
     assertThat(resolved).isInstanceOf(PklClassProperty::class.java)
-    assertThat(resolved.enclosingModule!!.moduleName).isEqualTo("k8s.api.apps.v1.Deployment")
+    assertThat(resolved.enclosingModule!!.name).isEqualTo("k8s.api.apps.v1.Deployment")
```
Comment (stackoverflow) [inline src/test/kotlin/org/pkl/lsp/GoToDefinitionPackagesTest.kt]: This assertion is failing. `name` only returns the last part, not the fully qualified name.

File: src/test/kotlin/org/pkl/lsp/actions/PklImplementMembersQuickFixTest.kt [hunk 4]
```diff
@@ -0,0 +1,465 @@
...
+      import "other.pkl" as bar
+
+      class Child extends foo.Base {
+        bar: bar2.Qux = TODO()
+      }
+      """
+        .trimIndent()
+    )
+  }
+
```
Comment (stackoverflow) [inline src/test/kotlin/org/pkl/lsp/actions/PklImplementMembersQuickFixTest.kt:422]: This is failing. Pretty sure is due to the auto-importer making assumptions about the order of imports.
```suggestion

 @Test
  fun `import type in different module, with chained alias conflicts`() {
    createPklFile(
      "foo.pkl",
      """
      abstract module foo

      import "bar.pkl"

      abstract class Base {
        abstract bar: bar.Qux
      }
      """
        .trimIndent(),
    )

    createPklFile(
      "bar.pkl",
      """
      class Qux
      """
        .trimIndent(),
    )

    createPklFile(
      "test.pkl",
      """
      import "aaa.pkl" as bar2
      import "foo.pkl"
      import "other.pkl" as bar

      class Child extends foo.Base
      """
        .trimIndent(),
    )
    implementMembersAndAssertResultEqualTo(
      """
      import "aaa.pkl" as bar2
      import "bar.pkl" as bar3
      import "foo.pkl"
      import "other.pkl" as bar

      class Child extends foo.Base {
        bar: bar3.Qux = TODO()
      }
      """
        .trimIndent()
    )
  }
  
```

File: src/test/kotlin/org/pkl/lsp/actions/PklImplementMembersQuickFixTest.kt [hunk 5]
```diff
@@ -0,0 +1,465 @@
...
+    implementMembersAndAssertResultEqualTo(
+      """
+      extends "foo.pkl"
+
+      bar: Int = TODO()
+      """
+        .trimIndent()
+    )
+  }
+
```
Comment (stackoverflow) [inline src/test/kotlin/org/pkl/lsp/actions/PklImplementMembersQuickFixTest.kt:260]: `lastMember` might return `null` if the module has no imports and properties/methods.
```suggestion

  @Test
  fun `quickfix implements module members without deleting trailing members`() {
    createPklFile(
      "foo.pkl",
      """
      abstract module foo

      abstract bar: Int
      """
        .trimIndent(),
    )

    createPklFile(
      "test.pkl",
      """
      extends "foo.pkl"

      class Bar {}
      """
        .trimIndent(),
    )
    implementMembersAndAssertResultEqualTo(
      """
      extends "foo.pkl"

      bar: Int = TODO()

      class Bar {}
      """
        .trimIndent()
    )
  }
  
```

File: src/main/kotlin/org/pkl/lsp/ast/Extensions.kt [hunk 6]
```diff
@@ -659,3 +661,184 @@ fun Appendable.renderParameterList(
...
+    }
+  }
+  return false
+}
+
+private fun appendNumericSuffix(name: String): String =
+  when (name.last()) {
+    in '0'..'9' -> {
+      val number = name.takeLastWhile { it in '0'..'9' }
+      name.substring(0, name.length - number.length) + (number.toInt() + 1)
```
Comment (stackoverflow) [inline src/main/kotlin/org/pkl/lsp/ast/Extensions.kt:737]: `toInt()` might overflow if the suffix is bigger than MAX_INT. Better to use a BigInt.
Yeah, it's a rare edge case, but we can't make assumptions about people's code.

**Review by stackoverflow — CHANGES_REQUESTED** [review]

File: src/test/kotlin/org/pkl/lsp/actions/PklImplementMembersQuickFixTest.kt (hunk 2 above)
Comment (bioball) [reply src/test/kotlin/org/pkl/lsp/actions/PklImplementMembersQuickFixTest.kt:162]: Great catch!

File: src/test/kotlin/org/pkl/lsp/actions/PklImplementMembersQuickFixTest.kt (hunk 1 above)
Comment (bioball) [reply src/test/kotlin/org/pkl/lsp/actions/PklImplementMembersQuickFixTest.kt:161]: Great catch!

File: src/test/kotlin/org/pkl/lsp/actions/PklImplementMembersQuickFixTest.kt (hunk 5 above)
Comment (bioball) [reply src/test/kotlin/org/pkl/lsp/actions/PklImplementMembersQuickFixTest.kt:260]: Nice find!

File: src/main/kotlin/org/pkl/lsp/ast/Extensions.kt (hunk 6 above)
Comment (bioball) [reply src/main/kotlin/org/pkl/lsp/ast/Extensions.kt:737]: This doesn't strike me as very important to handle. Likely, more things will blow up if you have that many imports. For example, our `Span` has `beginLine: Int`, `beginCol: Int`. Also, Pkl itself will probably blow up if trying to eval a module like this.

File: src/test/kotlin/org/pkl/lsp/actions/PklImplementMembersQuickFixTest.kt (hunk 4 above)
Comment (bioball) [reply src/test/kotlin/org/pkl/lsp/actions/PklImplementMembersQuickFixTest.kt:422]: Good catch!

File: src/test/kotlin/org/pkl/lsp/GoToDefinitionPackagesTest.kt (hunk 3 above)
Comment (bioball) [reply src/test/kotlin/org/pkl/lsp/GoToDefinitionPackagesTest.kt]: Ah, I re-introduced `moduleName`.

File: src/main/kotlin/org/pkl/lsp/ast/Extensions.kt (hunk 6 above)
Comment (stackoverflow) [reply src/main/kotlin/org/pkl/lsp/ast/Extensions.kt:737]: I was thinking more about having a single (or couple) imports with huge numbers as suffix, not that many imports. But I guess in that case this code would just pick a small number, right?

Comment (bioball) [reply src/main/kotlin/org/pkl/lsp/ast/Extensions.kt:737]: Yeah; just starts adding numeric suffixes until there's no conflict.

File: src/test/kotlin/org/pkl/lsp/actions/PklImplementMembersQuickFixTest.kt [hunk 7]
```diff
@@ -0,0 +1,575 @@
...
+        class Child extends Base2 {
+          name: String = TODO()
+
+          function greet(): String = TODO()
+        }
+        """
+          .trimIndent(),
+    )
+  }
+
```
Comment (stackoverflow) [inline src/test/kotlin/org/pkl/lsp/actions/PklImplementMembersQuickFixTest.kt:231]: This is failing (`name` is being re-implemented). I think this is due to `ExtendsClauseAnalyzer.checkAbstractMembers` and `PklImplementMembersQuickFix.getEdits` having different rules to wheter/which members get inserted.
```suggestion

@Test
  fun `quickfix only implements members not already declared`() {
    checkQuickFix(
      before =
        """
        abstract module Foo

        abstract class Base {
          abstract name: String
          abstract age: Int
        }
        class Child extends Base {
          name: String = "Bob"
        }
        """
          .trimIndent(),
      after =
        """
        abstract module Foo

        abstract class Base {
          abstract name: String
          abstract age: Int
        }
        class Child extends Base {
          name: String = "Bob"

          age: Int = TODO()
        }
        """
          .trimIndent(),
    )
  }
```

File: src/main/kotlin/org/pkl/lsp/analyzers/ExtendsClauseAnalyzer.kt [hunk 8]
```diff
@@ -0,0 +1,276 @@
...
+            PklAddModifierQuickFix(
+              resolved,
+              ErrorMessages.create("addModuleModifier", moduleName, "abstract"),
+              TokenType.ABSTRACT,
+            )
+        }
+      }
+    }
+    val module = clause.enclosingModule ?: return
+    checkParentClassDef(clause, module, project.pklBaseModule, holder, context)
```
Comment (stackoverflow) [inline src/main/kotlin/org/pkl/lsp/analyzers/ExtendsClauseAnalyzer.kt:86]: Not entirely sure, but this might generate a stack overflow for a circular inheritance (`effectiveParentProperties` -> `parentTypeDef` -> `cache`). Might be a good idea to create a test just to be sure.

**Review by stackoverflow — APPROVED** [review]
One issue left. Approving to unblock.

File: src/test/kotlin/org/pkl/lsp/actions/PklImplementMembersQuickFixTest.kt (hunk 7 above)
Comment (bioball) [reply src/test/kotlin/org/pkl/lsp/actions/PklImplementMembersQuickFixTest.kt:231]: Ah, whoops, that was a bug I introduced in my latest commit. Fixed!

File: src/main/kotlin/org/pkl/lsp/analyzers/ExtendsClauseAnalyzer.kt (hunk 8 above)
Comment (bioball) [reply src/main/kotlin/org/pkl/lsp/analyzers/ExtendsClauseAnalyzer.kt:86]: This does stackoverflow! But I think this is an orthogonal bug in our resolving logic; merging this for now and we can fix in a follow-up.
